docs: document subscriptions/listen and CacheableResult in draft learn and client pages (#3068)
modelcontextprotocol/modelcontextprotocol: docs/docs/draft/learn/architecture.mdx

**Server Capabilities**:

- `"tools": {"listChanged": true}` - The server supports the tools primitive and can honor a `toolsListChanged` filter in `subscriptions/listen`. Clients that request this filter receive `notifications/tools/list_changed` when the tool list changes.

> - `"resources": {}` - The server also supports the resources primitive (can handle `resources/list` and `resources/read` methods)

Reviewer: Can we link to the subscription pages?

            "required": ["location"]
          }
        }
      ],
      "ttlMs": 300000,
      "cacheScope": "public"
    }
  }
  ```
</CodeGroup>

#### Understanding the Tool Discovery Request

> The `tools/list` request carries only the standard `_meta` fields that identify the client on every request, with no tool-specific parameters.

Reviewer: I think in spirit the original text "The `tools/list` request is simple, containing no parameters." is fine.

- **`description`**: Detailed explanation of what the tool does and when to use it
- **`inputSchema`**: A JSON Schema that defines the expected input parameters, enabling type validation and providing clear documentation about required and optional parameters

The result is marked `"resultType": "complete"` and carries two caching fields that complete list and read results must include:

- **`ttlMs`**: A freshness hint in milliseconds. The client may treat the response as fresh for this long before re-fetching. Here `300000` means the tool list can be cached for five minutes.
- **`cacheScope`**: Indicates who may cache the response. `"public"` responses contain no user-specific data and may be served from shared caches. `"private"` responses may only be reused within the same authorization context.

> 

Reviewer: Having cache handling peppered throughout is not great. We should have a page or a section about caching and refer to that instead of repeating parts of the caching behaviour.

#### Understanding Tool List Change Notifications

When the server's available tools change (for example, when new functionality becomes available, existing tools are modified, or tools become temporarily unavailable), the server notifies clients that opted in:

```json Notification
{
  "jsonrpc": "2.0",
  "method": "notifications/tools/list_changed",
  "params": {
    "_meta": {
      "io.modelcontextprotocol/subscriptionId": 4
    }
  }
}
```

#### Key Features of MCP Notifications

1. **No Response Required**: Notice there's no `id` field in the notification. This follows JSON-RPC 2.0 notification semantics where no response is expected or sent.

2. **Opt-In Based**: This notification is only sent to clients that requested `"toolsListChanged": true` in their `subscriptions/listen` filter.

3. **Subscription-ID Tagging**: Every notification on the stream carries `io.modelcontextprotocol/subscriptionId` in `_meta`. The value is the JSON-RPC ID of the `subscriptions/listen` request that opened the stream (`4` in this example), so clients can correlate each notification with the subscription that produced it.

> 4. **Event-Driven**: The server decides when to send notifications based on internal state changes, making MCP connections dynamic and responsive.
> 

Reviewer 2: ```suggestion
4. **Event-Driven**: The server decides when to send notifications based on internal state changes, making MCP connections dynamic and responsive.
5. **Best Effort**: There are no guarantees that every notification will be sent/received, particularly across transport reconnects. Clients should also rely on polling to preserve freshness of results.

```

<Step title="Real-time Updates (Notifications)">
MCP supports real-time notifications that enable servers to inform clients about changes without being explicitly requested. This demonstrates the notification system, a key feature that keeps MCP connections synchronized and responsive.

> #### Opting In with `subscriptions/listen`

Reviewer: "Subscribing to changes"

Change notifications are opt-in. To receive them, the client opens a long-lived notification stream by sending a `subscriptions/listen` request with a `notifications` filter naming the event types it wants. Here the client asks for tool list changes:

```json Request
{
  "jsonrpc": "2.0",
  "id": 4,
  "method": "subscriptions/listen",
  "params": {
    "_meta": {
      "io.modelcontextprotocol/protocolVersion": "2026-07-28",
      "io.modelcontextprotocol/clientInfo": {
        "name": "example-client",
        "version": "1.0.0"
      },
      "io.modelcontextprotocol/clientCapabilities": {}
    },
    "notifications": {
      "toolsListChanged": true
    }
  }
}
```

> Every client request carries the `io.modelcontextprotocol/protocolVersion`, `clientInfo`, and `clientCapabilities` fields in `_meta`, so the server can identify the client without relying on connection state.

Reviewer 3: #3002 on `main` made `io.modelcontextprotocol/clientInfo` a SHOULD (the per-request `_meta` table on the [draft Overview page](https://modelcontextprotocol.io/specification/draft/basic) marks it not required), so only `protocolVersion` and `clientCapabilities` are on every request.

```suggestion
Every client request carries the `io.modelcontextprotocol/protocolVersion` and `io.modelcontextprotocol/clientCapabilities` fields in `_meta`, and normally `io.modelcontextprotocol/clientInfo` as well, so the server can identify the client without relying on connection state.
```

> The server acknowledges the filter with `notifications/subscriptions/acknowledged` as the first message on the stream, then delivers matching notifications as changes occur. The server never sends notification types the client did not request.

Reviewer 3: The [Subscriptions page](https://modelcontextprotocol.io/specification/draft/basic/patterns/subscriptions) defines the ack ordering per subscription ID rather than per stream ("the first message carrying the subscription's ID in `_meta`"), and on stdio other subscriptions' messages **MAY** be interleaved before it, so "first message on the stream" overstates the guarantee.

```suggestion
The server acknowledges the filter with `notifications/subscriptions/acknowledged`, which is the first message carrying that subscription's ID in `_meta`, then delivers matching notifications as changes occur. The server never sends notification types the client did not request.
```

#### Understanding Tool List Change Notifications

> When the server's available tools change—such as when new functionality becomes available, existing tools are modified, or tools become temporarily unavailable—the server notifies clients that opted in:

Reviewer 3: Em dashes.

```suggestion
When the server's available tools change (for example, when new functionality becomes available, existing tools are modified, or tools become temporarily unavailable), the server notifies clients that opted in:
```
